src/__generated__/ShippingAndPaymentReview_order.graphql.ts

```diff
@@ -1,17 +1,18 @@
 /* tslint:disable */
 
 import { ConcreteFragment } from "relay-runtime";
-export type OrderFulfillmentType = "PICKUP" | "SHIP" | "%future added value";
 declare const _ShippingAndPaymentReview_order$ref: unique symbol;
 export type ShippingAndPaymentReview_order$ref = typeof _ShippingAndPaymentReview_order$ref;
 export type ShippingAndPaymentReview_order = {
-    readonly fulfillmentType: OrderFulfillmentType | null;
-    readonly shippingName: string | null;
-    readonly shippingAddressLine1: string | null;
-    readonly shippingAddressLine2: string | null;
-    readonly shippingCity: string | null;
-    readonly shippingPostalCode: string | null;
-    readonly shippingRegion: string | null;
+    readonly requestedFulfillment: ({
```

Comment:
@alloy would it be possible to get named types that represent either side of the union here? If you bring in `__typename` automatically for unions it would even be possible to generate predicate functions in this file too, e.g. `isShip` and `isPickup`

Reply:
That is somewhat how it’s supposed to work and why [I requested](https://github.com/artsy/reaction/pull/1193#discussion_r213549330) the change to select `__typename` instead.

For instance, take [this union selection](https://github.com/artsy/emission/blob/b743779dff6480deedfe61a51dfcbc7fa9324a9b/src/lib/Components/Inbox/Conversations/ConversationSnippet.tsx#L189-L208) as an example. The [emitted typings](https://github.com/artsy/emission/blob/master/src/__generated__/ConversationSnippet_conversation.graphql.ts#L13-L35) for it are a union with each their own `__typename: "TypeName"` field. This is done so that TS can automatically narrow down the type based on the `__typename` value, which you can see in action [here](https://github.com/artsy/emission/blob/b743779dff6480deedfe61a51dfcbc7fa9324a9b/src/lib/Components/Inbox/Conversations/ConversationSnippet.tsx#L145-L149) where TS knows that if `__typename === "Artwork"` the `item` can only be the `Artwork` part of the union etc.

Reply:
Yes we do, this is the repo https://github.com/relay-tools/relay-compiler-language-typescript. I actually filed an issue just this morning about emitting named types for selections https://github.com/relay-tools/relay-compiler-language-typescript/issues/64.

Could you expand a bit on why we’d need it in this case?

Reply:
to be able to have sub-components or methods that don't care about the union type and don't need to specify their input types independently of the generated types. e.g.

```
class ShippingInformation {
  renderAddress({name, addressLine1, ...}: Ship) {
     return ...
  }
  render() {
    const {requestedFulfilment, shippingOrigin} = this.props.order

    return isShip(requestedFulfilment)
       ? renderAddress(requestedFulfilment)
       : "Pick up (${shippingOrigin})"
  }
}

Reply:
Gotcha.

FWIW Sub-components can work _today_ if you’d give them their own container and fragment.

Reply:
Or, as you [pointed out](
https://github.com/relay-tools/relay-compiler-language-typescript/issues/64#issuecomment-41692849)3, by using the `Extract` type 👍